hack/upgrade-test.sh

```diff
@@ -0,0 +1,282 @@
+readonly SOURCE_BRANCH="main"
+readonly CONFIG_REPO="gitops-test"
+readonly CONFIG_BRANCH="main"
+readonly CONFIG_COMMIT_MESSAGE="Update config"
+
+main() {
+  ./hack/setup.sh cluster cartographer-latest example-dependencies
+
+  port=$(available_port)
+  setup_git_server "$port"
+
+  source_dir="$(mktemp -d)"
+  setup_source_repo "$source_dir" "$port"
+  setup_source_to_gitops
+
+  config_dir="$(mktemp -d)"
+  test_source_to_gitops "$config_dir" "$port"
+
+  setup_gitops_to_app
+  test_gitops_to_app
+
+  log "deploying cartographer from main"
+  ./hack/setup.sh cartographer
+  verify_new_carto_deployed
+
+  update_source "$source_dir" "$port"
+  wait_for_new_config_commit "$config_dir"
+  wait_for_knative_deployment_update
+}
+
+setup_git_server() {
+  local port=$1
+
+  log "setting up git server"
+
+  kubectl apply -f ./hack/git-server.yaml
+
+  until [[ $(kubectl get deployment git-server -o json | jq '.status.readyReplicas') == 1 ]]; do
+    log "waiting for git-server deployment"
+    sleep 10
+  done
+
+  (kubectl port-forward service/git-server $port:80 &) >/dev/null
+  sleep 5
+}
+
+setup_source_repo() {
+  local source_dir=$1
+  local port=$2
+
+  log "setting up source repo"
+
+  pushd "$source_dir"
+    git clone "http://localhost:$port/$SOURCE_REPO.git"
+    pushd "$SOURCE_REPO"
+      git pull https://github.com/kontinue/hello-world.git
+      if [[ $(git branch --show-current) != "$SOURCE_BRANCH" ]]; then
+        git checkout -b $SOURCE_BRANCH
+      fi
+      git push origin $SOURCE_BRANCH
+    popd
+  popd
+}
+
+setup_source_to_gitops() {
+  log "setting up source to gitops"
+
+  kapp deploy --yes -a "setup-example-$TEST_NAME" \
+    -f <(ytt --ignore-unknown-comments \
+      -f "$DIR/../examples/shared" \
+      -f "$DIR/../examples/$TEST_NAME/values.yaml" \
+      --data-value registry.server="$REGISTRY" \
+      --data-value registry.username=admin \
+      --data-value registry.password=admin \
+      --data-value image_prefix="$REGISTRY/example-$TEST_NAME-")
+
+  kapp deploy --yes -a "example-$TEST_NAME" \
+    -f <(ytt --ignore-unknown-comments \
+      -f "$DIR/../examples/$TEST_NAME" \
+      --data-value registry.server="$REGISTRY" \
+      --data-value registry.username=admin \
+      --data-value registry.password=admin \
+      --data-value workload_name="$TEST_NAME" \
+      --data-value image_prefix="$REGISTRY/example-$TEST_NAME-" \
+      --data-value source_repo.url="http://$GIT_SERVER/$SOURCE_REPO.git" \
+      --data-value source_repo.branch="$SOURCE_BRANCH" \
+      --data-value git_repository="http://$GIT_SERVER/$CONFIG_REPO.git" \
+      --data-value git_branch="$CONFIG_BRANCH" \
+      --data-value git_user_name="gitops-user" \
+      --data-value git_user_email="gitops-user@example.com" \
+      --data-value git_commit_message="$CONFIG_COMMIT_MESSAGE")
+}
+
+test_source_to_gitops() {
+  local config_dir=$1
+  local port=$2
+
+  log "testing source to gitops"
+
+  SUCCESS=false
```

Comment:
I think these upper case variables could be not only lowercase, but also marked `local` as they're supposed to only be available inside the function - iirc, we've been reserving upper case ones for those that have definitions at the global scope usually with some default that can be overwritten via environment variables